app/public/consoleWindow.js

```diff
@@ -0,0 +1,92 @@
+const consoleWindowElement = document.getElementById("consoleWindow");
+
+const lineHolderElement = document.createElement("div");
+lineHolderElement.className = "lineHolder";
+consoleWindowElement.appendChild(lineHolderElement);
+
+const scrollElement = consoleWindowElement;
+
+const snapToBottomTolerance = 50; //When within this many pixels from the bottom of the console, new messages will cause the scroll to snap to the bottom
+let modes = ["public", "private", "important", "critical"];
+let seconds = 60*4; //for testing
+
+//Expects an object with 3 keys: time, message, classString
+function AddConsoleLine(lineObj){
+    let {time, message, classString} = lineObj;
+
+    let stringMessage = "";
+
+    if(time !== undefined){
+
+        let timeString = new Date(time * 1000).toISOString().substring(14, 19);
+        stringMessage += `${timeString} `;
+    }
+
+    stringMessage += "> "+message;
+    const newElem = AddConsoleLineString(stringMessage);
+    newElem.className = classString;
+}
+
+//Makes and returns a line element as a child of the lines parent element, setting its text content
+function AddConsoleLineString(lineString){
+    let scrolledToBottom = IsScrolledToBottom();
+    let lineElement = document.createElement("div");
+    lineElement.textContent = lineString;
+    lineHolderElement.appendChild(lineElement);
+
+    if(scrolledToBottom){
```

Comment:
nitpick, but we could have if(IsScrolledToBottom()) here and not have to store it into the variable

Reply:
I had tried that at first, but I actually need to check and cache whether it is scrolled to the bottom _before_ adding the new element. Otherwise, IsScrolledToBottom() always returns false. The idea is that if the user was  fully scrolled down, they should remain scrolled down after the new line is appended.